### ISSUE TITLE
Exported vars should not be assumed constant

### ISSUE THREAD

**Issue by tadeegan** [google/closure-compiler#2251]
If I write the following code I get an error complaining that "XX == null" is always true. If I export XX then it is possible someone outside of the compile is setting it to not null right?

```
/** @export */
var XX = null;

function hello() {
  if (XX == null) {
    console.log("whats up!");
  }
}
```

Specifically, the error I am seeing is:
```
WARNING - condition always evaluates to true
left : null
right: null
  if (XX == null) {
      ^^^^^^^^^^

0 error(s), 1 warning(s), 100.0% typed

```

**Comment by ChadKillingsworth** [comment]
I believe it's being assumed constant because the coding convention assumes upper case variables are constants. In any case, exported variables cannot be changed by external code. 

```js
var foo = 'foo'

// the export is an alias
window.bar = foo;

// Change in external code
window.bar = 'bar';

console.log(foo); //logs 'foo'
```

**Comment by Dominator008** [comment]
This particular check has been a source of false positives in OTI. AFAIK NTI doesn't have this check.

**Comment by concavelenz** [comment]
It isn't the check that is the problem in the OTI but the inferred type is sometimes too tight, not accounting for the assignments in local scopes.  Declaring the type always fixes the problem is I recall correctly.

**Comment by tadeegan** [comment]
I think Chad is right here.  I was thinking about exports kind of like externs but they are quite different. External code can't actually assign 'XX' anymore as code internal to the compile may have renamed it and just aliased like it this:

`XX = <some compiled name>`

Seems like it is a bad idea to ever export a value that can change.  Should exporting be restricted to consts, classes, and class methods? Seems like if you are just giving an alias to a certain value, then that internal value may change over time and you will never see the update. For example:

```
var foo = 'foo'
// the export is an alias
window.bar = foo;
// value changes internally
window.setTimeout(1000, function() {foo = 'asdf'});

// Read export in external code
console.log(window.bar); // always logs 'foo'!
```

Also, @ChadKillingsworth the upper case does not affect the outcome of this:
https://closure-compiler-debugger.appspot.com/#input0%3D%252F**%2520%2540export%2520*%252F%250Avar%2520cc%2520%253D%2520null%253B%250A%250Afunction%2520hello()%2520%257B%250A%2520%2520if%2520(cc%2520%253D%253D%2520null)%2520%257B%250A%2520%2520%2520%2520console.log(%2522whats%2520up!%2522)%253B%250A%2520%2520%257D%250A%257D%26input1%26conformanceConfig%26externs%3Dvar%2520goog%253B%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CLOSURE_PASS%3D1%26COLLAPSE_PROPERTIES%3D1%26DEVIRTUALIZE_PROTOTYPE_METHODS%3D1%26DISAMBIGUATE_PROPERTIES%3D1%26GENERATE_EXPORTS%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1%26VARIABLE_RENAMING%3D1
